### ISSUE TITLE
Multiprocessing with multithreading

### ISSUE THREAD

**Issue by ochedru** [prometheus/client_python#431]
Our main process is multithreaded and forks child processes. I acknowledge this can be a recipe for disaster when `threading.Lock` instances are acquired all over the place.
No surprise, I encountered this exact issue with Prometheus metrics with multiprocessing mode enabled:
```
Current thread 0x00007fd4ec3ba700 (most recent call first):
  File "/usr/lib/python3.7/site-packages/prometheus_client/values.py", line 86 in set
  File "/usr/lib/python3.7/site-packages/prometheus_client/metrics.py", line 344 in set
```
The Prometheus gauge is created in the main process. It owns a `threading.Lock` instance which obviously can deadlock after fork. There is a nice comment about it:
```
def MultiProcessValue(_pidFunc=os.getpid):
    files = {}
    values = []
    pid = {'value': _pidFunc()}
    # Use a single global lock when in multi-processing mode
    # as we presume this means there is no threading going on.
    # This avoids the need to also have mutexes in __MmapDict.
    lock = Lock()
```
So maybe the documentation should include "no multithreading" as a limitation for the multiprocessing mode. Anyway, I am trying to find a workaround:
- Create new metrics with the same name in the child process just after fork. Will this work with multiprocessing mode enabled?
- Could the Prometheus client code be changed to use a `multiprocessing.Lock` instead of a `threading.Lock`?

Thank you for reading!

**Comment by brian-brazil** [comment]
We presume there's no threading going on in terms of performance tradeoffs, it should still handle this care correctly.

The issue here sounds more like forking and threading, which I'm not sure we can do much about as it's a general problem.

> Could the Prometheus client code be changed to use a multiprocessing.Lock instead of a threading.Lock?

Would that still handle cases where the forking is not related to the `multiprocessing` module?

**Comment by ochedru** [comment]
The `multiprocessing.Lock` is implemented using an OS semaphore, which handle is inherited by the forked child process. So this should work whatever the forking method.

**Comment by brian-brazil** [comment]
We can't presume that our code runs in the parent process though, in some setups it only ever runs in the children.

**Comment by ochedru** [comment]
In this case, you get new lock instances in the child process and it is fine. The purpose of these locks is to prevent concurrent metrics updates from within the same process, or did I miss something?

**Comment by brian-brazil** [comment]
That'd create a lock across child processes though, which is a bit inefficient.

**Comment by ochedru** [comment]
Indeed. We could leverage `os.register_at_fork()` to replace (potentially locked) metric locks by fresh new locks on all instances immediately after the forked process starts.

**Comment by brian-brazil** [comment]
That's only available in 3.7.

**Comment by ochedru** [comment]
OK. FYI, to make it work in my situation, I ended up replacing at runtime the `_lock` attribute of my metrics with a `multiprocessing.Lock` instance. I guess we can close this discussion.
Thank you for your time!
